File: grokking-algorithm/binary-search/1.py
def binary_search(arr, item):
  low = 0                                   # ①
  high = len(arr) - 1                       # ①

  while low <= high:                        # ②
    mid = (low + high) // 2                 # ③
    guess = arr[mid]
    if guess == item:                       # ④
      return mid
    elif guess > item:                      # ⑤
      high = mid - 1
    else:                                   # ⑥
      low = mid + 1
  return None                               # ⑦

my_list = [1, 3, 5, 7, 9]                   # ⑧

print(binary_search(my_list, 3)) # => 1       ⑨
print(binary_search(my_list, -1)) # => None   ⑩

File: grokking-algorithm/binary-search/1.ts
export {};

const binarySearch = (list: number[], item: number): number | null => {
	let left = 0;
	let right = list.length - 1;

	while (left <= right) {
		const middle = Math.floor((left + right) / 2);
		const guess = list[middle];

		if (guess === item) return middle;

		if (guess > item) {
			right = middle - 1;
		} else {
			left = middle + 1;
		}
	}

	return null;
};

const myList = [1, 3, 5, 7, 9];

console.log(binarySearch(myList, 3)); // => 1       ⑨
console.log(binarySearch(myList, -1)); // => null   ⑩
